### DISCUSSION TITLE
How to deploy Developer Blog tutorial to Vercel?

### DISCUSSION THREAD

**Discussion by DennisKo** [remix-run/remix#1492] [Q&A] [answered]
Title says it already. What would be the next steps after completing the [Dev blog tutorial](https://remix.run/docs/en/v1/tutorials/blog) if you want to deploy to Vercel?

I am stuck at https://github.com/remix-run/remix/issues/564 where the `posts` dir cannot be found. Vercel makes it really hard to debug that as the function logs and the output view seem to show different things.

**Comment by silvenon** [comment, marked as answer]
I noticed that Vercel published `public`, so I put my posts in `public/posts`, you can check the [repo](https://github.com/silvenon/silvenon.com/tree/21dac4835a913678c7fdd94b7fad21803a88ceac). The full story is that I keep them in `app/posts` to retain live reload while editing posts, then I copy them to `public/posts` before deployment (ignoring `public/posts` in git). Kind of clumsy but it… well, I can't say that it works yet because I'm having trouble using esbuild in Vercel, but it probably would've worked. 😄 

I asked for some guidance in #1501 about whether there's a better way to do it.

**Reply by DennisKo** [reply]
Hmm yes thanks, I am now one step further: I put my posts (*.md files) into `public/posts` and on the vercel deployment I now get this structure:  
![Screenshot 2022-01-14 at 17 36 53](https://user-images.githubusercontent.com/9072277/149551653-1cdf67df-fe9f-49c5-83ee-8394edb1808b.png)

When I log out my postsPath `path.join(__dirname, "..", "posts")` I get: `/var/task/output/server/pages/api/posts`. I just dont get how this path matches the above dir structure in any way? Where is this been logged?

**Reply by silvenon** [reply]
The code where you're logging from is bundled inside `api/_build`, so it sounds like in production you just need to go up twice: `path.join(__dirname, "../..", "posts")`

It's still a little confusing to me how to make the path be the same both in development and in production. Perhaps keep `posts` in the project root, then before deployment copy it to `public/posts`, that way the path might stay the same both in development and in production.

If you're confused about the `/var/task/output/server/pages` part, I guess that's just where Vercel deployed the code.

**Reply by silvenon** [reply]
Another thing to try is to keep the posts in `api/posts`, that way there should be no duplication required 🤞  That way your code `path.join(__dirname, "..", "posts")` can remain the same.

**Reply by DennisKo** [reply]
I think I am missing something. I logged all directories in the deployment now - there is not `posts` folder. Although it clearly shows up in the source view of the vercel deployment

Btw. I have
```
const postsPath =
  process.env.NODE_ENV === "development"
    ? path.join(__dirname, "../../public", "posts")
    : path.join(__dirname, "../../..", "posts");
```

to make it work in dev and prod (once it actually works...) without manually moving stuff around.

**Reply by silvenon** [reply]
> all directories in the deployment

What do you mean by "deployment" here? And what did you log exactly?

I think putting posts in `api/posts` might be best, because `api/__build/index.js` is the file that is reading them, both in development and in production.

I wish Remix had a clear recommendation for this, though, I think it would be totally fine for you to post this as an issue in Remix, as the documentation is not complete for this.

**Reply by DennisKo** [reply]
I mean by that logging in the serverless function. I log out everything in a dir and tried different levels. If I put the posts into `api/posts` I cant even find them deployed anywhere ...

**Reply by silvenon** [reply]
I actually tried it now instead of just guessing, and you're right, if I put it in `api/posts` it doesn't appear in the deployment's output, and if I put it to `public/posts` it's present in the output but `api/index.js` still cannot access it for some reason.

I guess this means that I have the same problem with my blog as well, I just didn't notice because I have another problem that's happening sooner, before trying to read the blog posts.

I encourage you to create an issue.

**Reply by DennisKo** [reply]
Let me tinker with it a bit more today. TBH I am not totally sure how these serverless environments are handled (by vercel in that case). To me it seems that the actual deployment (assets, html, etc.) is running in a different environment than the function. I will have to read up on that...good luck with your blog! ;)

**Reply by silvenon** [reply]
Yeah, this is very new to me too. Thanks, good luck as well, and post back here if you find the solution. 👍

**Reply by frendyguo-zz** [reply]
Holy shit, this actually works. I have been dealing with the same problem for 2 nights. You're the man @silvenon!

**Reply by silvenon** [reply]
I'm very glad to hear that it works! And also confused because two of us couldn't make it work 😆 You can maybe share some details of what you did exactly.

**Reply by frendyguo-zz** [reply]
Wait, that's weird, I actually followed what you did in your repo, and it's clearly working, no?

Anyway, this is how I did it, https://github.com/frendyguo/remix.frendyguo.me
And this is the result https://remix.frendyguo.me/

Works locally and in production without checking the environment

**Reply by silvenon** [reply]
🤔  So the step that worked was copying `app/posts` to `public/posts`, are you sure? It doesn't work without it?

Btw, be careful, you committed `.vercel` to version control, but the readme inside says that you shouldn't share it with anyone, so you might want to delete it from the history entirely.

**Reply by frendyguo-zz** [reply]
> Btw, be careful, you committed .vercel to version control, but the readme inside says that you shouldn't share it with anyone, so you might want to delete it from the history entirely.

Oh wow, thanks for reminding me that.

**Reply by frendyguo-zz** [reply]
> 🤔 So the step that worked was copying app/posts to public/posts, are you sure? It doesn't work without it?

I just tried this, not copying `app/posts` to `public/posts` still works.

The magic seems to be in this path `${__dirname}/../../app`, which points to the deployed app directory.

**Reply by silvenon** [reply]
Awesome, that's the one thing we haven't tried! What was confusing to me was that `app` directory isn't visible in the deployment output, but it's apparently accessible to the serverless function. One of these days I'm going to figure out why. 😆 

Thanks so much for your comments!

**Comment by Adbib** [comment]
make your Posts function like this I solve this problem and its work for me on vercel

```js

export async function getPosts() {
  const postsPath = await fs.readdir(
    `${__dirname}/../../app/posts`,
    { withFileTypes: true }
  );

  const posts = await Promise.all(
    postsPath
      .map(async (dirent) => {
        const file = await fs.readFile(
          path.join(`${__dirname}/../../app/posts`, dirent.name)
        );
        const { attributes } = parseFrontMatter(file.toString());
        return {
          slug: dirent.name.replace(/\.mdx/, ""),
          title: attributes.title,
        };
      })
  );

  return posts;
}
```

**Reply by frendyguo-zz** [reply]
I'm actually not sure why this `${__dirname}/../../app` work in vercel function, is this where vercel actually host all the assets?

**Reply by JoaoCnh** [reply]
Is this still working? If I inspect `/var/task/app` (given that Remix is running in `/var/task/api`) and there is nothing. Not even in `/var/app`. I can't access my public files to read 🤔
